zerver/views/message_summary.py

```diff
@@ -0,0 +1,131 @@
+import json
+from typing import Any
+
+import litellm
+from django.conf import settings
+from django.http import HttpRequest, HttpResponse
+from django.utils.timezone import now as timezone_now
+from pydantic import Json
+
+from analytics.lib.counts import COUNT_STATS, do_increment_logging_stat
+from zerver.lib.narrow import NarrowParameter
+from zerver.lib.response import json_success
+from zerver.lib.typed_endpoint import typed_endpoint
+from zerver.models import UserProfile
+from zerver.views.message_fetch import get_messages_backend
+
+# Maximum number of messages that can be summarized in a single request.
+MAX_MESSAGES_SUMMARIZED = 100
+OUTPUT_COST_PER_TOKEN = 5
+INPUT_COST_PER_TOKEN = 1
+
+
+def format_conversation(zulip_messages: list[dict[str, Any]]) -> str:
+    # Note: Including timestamps seems to have no impact; including reactions
+    # makes the results worse.
+
+    zulip_messages_list = [
+        {"sender": message["sender_full_name"], "content": message["content"]}
+        for message in zulip_messages
+    ]
+    return json.dumps(zulip_messages_list)
+
+
+def make_message(content: str, role: str = "user") -> dict[str, str]:
+    return {"content": content, "role": role}
+
+
+def get_max_summary_length(conversation_length: int) -> int:
+    return min(6, 4 + int((conversation_length - 10) / 10))
```

Comment:
A comment explaining the logic behind this algorithm would be helpful.

Reply:
I copied it from the `python-zulip-api` code which I guess @alya wrote. I don't know the reasoning here. But, we should also take the message content length into account here.